### PR TITLE
[RFC][FlowBundle] Use Storage component

### PR DESCRIPTION
**Pull request by adamelso**
| Q             | A
| ------------- | ---
| Bug fix?      | -
| New feature?  | yes
| BC breaks?    | Application: no; Bundle: yes
| Deprecations? | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

At first, I thought FlowBundle was using Storage, but it actually has its own storage classes and interfaces.
The difference appears to be that FlowBundle accesses session variables from a custom bag.

Is it worth merging the two somehow to reduce code duplication and to make it easier to add custom storage? This might also give way to providing a fix for #3538

### CODE REVIEW

File: src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php [hunk 1]
```diff
@@ -46,7 +46,7 @@ function it_initializes(
     ) {
         $process->getScenarioAlias()->shouldBeCalled();
         $storage->initialize(Argument::type('string'))->shouldBeCalled();
-        $process->getOrderedSteps()->shouldBeCalled()->willReturn(array($previousStep, $currentStep, $nextStep));
+        $process->getOrderedSteps()->shouldBeCalled()->willReturn([$previousStep, $currentStep, $nextStep]);
```
Comment (stloyd) [inline src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php]: As you change this file anyway, please remove all calls to `shouldBeCalled()` if there are calls to the `willReturn()` like on this line.

**Comment by pjedrzejewski** [comment]
Definitely makes sense.

I know that @peteward will kill me for saying this but I have this crazy idea laying around for some time already... Maybe I will illustrate it with some routing:

Rough idea of routing:

```
sylius_checkout_addresing:
    path: /addressing
    methods: [GET, PUT]
    defaults:
        _controller: sylius.controller.cart:updateAction
        _sylius:
            template: SyliusShopBundle:Checkout:addressing.html.twig
            form: sylius_checkout_addressing
            event: checkout_addressing
            repository:
                method: findOneByCustomerAndCheckoutState
                arguments: ["expr:service('sylius.context.customer').get()", 'started']
            state_machine:
                transition: address
                graph: sylius_cart_checkout
            redirect: sylius_checkout_shipping
```

Before anyone calls me crazy :D I am not entirely convinced too but using tested and simple actions instead of custom controllers gives us less points for possible bugs, with flow bundle, custom update logic etc. All is handled by our good old RC above.

Cons:
* Too much logic in routing?

Pros:
* All FlowSteps are basically this right now, updating the order using appropriate form, dispatching some events, next, that's it. In #3918 we add state machine and that perfectly fits into this scheme.
* Works for both HTML rendering and API. We would have an unified way of handling the checkout and would not need to worry about differences between api / html access that much.
* Easy to customize, probably easier than overriding whole Step class.
* Less code is better. :D
* There are other use cases for introducing state machine into updateAction, for example:

Shipping the shipment. Even if we drop this idea for checkout, I feel like it makes sense for many other actions, shipping the shipment is just an example. Shipping the shipment means executing transition "ship" on shipment and optionally setting tracking code (or whatever you want if you customize the form), so that could be perfectly realized by:

```
sylius_admin_shipment_ship:
    path: /{id}/ship
    defaults:
        _controller: sylius.controller.shipment:updateAction
        _sylius:
            template: SyliusAdminBundle:Shipment:ship.html.twig
            form: sylius_shipment_ship
            event: ship
            state_machine:
                transition: ship
```

No worries, this is not decided yet but I believe that could be used for many other transitions which need to edit something as well. Let me know! :dancer:

File: src/Sylius/Bundle/FlowBundle/spec/Storage/SessionStorageSpec.php [hunk 2]
```diff
@@ -30,42 +30,55 @@ function let(SessionInterface $session, SessionFlowsBag $bag)
     function it_is_initializable()
     {
         $this->shouldHaveType('Sylius\Bundle\FlowBundle\Storage\SessionStorage');
-        $this->shouldHaveType('Sylius\Bundle\FlowBundle\Storage\AbstractStorage');
+    }
+
+    function it_is_a_Sylius_storage()
+    {
+        $this->shouldImplement('Sylius\Component\Storage\StorageInterface');
```
Comment (stloyd) [inline src/Sylius/Bundle/FlowBundle/spec/Storage/SessionStorageSpec.php]: `StorageInterface::class` same below.

**Comment by peteward** [comment]
fix provinces first!! magic code later.

**Comment by pjedrzejewski** [comment]
@peteward I have commented in the relevant issue. Checkout code is quite important as well,  that's why I am throwing an idea here.

File: src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php (hunk 1 above)
Comment (bendavies) [inline src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php]: @stloyd does `willReturn` imply `shouldBeCalled`?

Comment (adamelso) [inline src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php]: @stloyd :+1: 

@bendavies It's part of the [Command-Query Separation principle](http://martinfowler.com/bliki/CommandQuerySeparation.html).

Basically, methods should either:

  * Do something (_Command_, `shouldBeCalled()`)
  * Return something (_Query_, `willReturn()`)

but never both!

However, for methods that don't adhere to this principle, you'll need to do `->shouldBeCalled()->willReturn()`, but this is not the case here.

Comment (bendavies) [inline src/Sylius/Bundle/FlowBundle/spec/Process/Context/ProcessContextSpec.php]: @adamelso thanks!

**Comment by adamelso** [comment]
FlowBundle is currently not deprecated yet, but will be soon. Closing. (and last PR to close, the remaining PR I'll keep open :).

**Pull request closed without merging**